### DISCUSSION TITLE
[Feature request] Support for SRI

### DISCUSSION THREAD

**Discussion by Exelord** [vercel/next.js#23951] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

[Subresource Integrity](https://www.w3.org/TR/SRI/) (SRI) is a security feature that enables browsers to verify that files they fetch (for example, from a CDN) are delivered without unexpected manipulation.

## Describe the solution you'd like

Implement webpack plugin (https://www.npmjs.com/package/webpack-subresource-integrity) as default, or under a flag just for production environment.

Thanks to it every external resource should have generated `integrity` hash and specified `crossorigin` policy. Example:
```html
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css" integrity="sha384-Gn5384xqQ1aoWXA+058RXPxPg6fy4IWvTNh0E263XmFcJlSAwiGgFAW/dAiS6JXm" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/js/bootstrap.min.js" integrity="sha384-JZR6Spejh4U02d8jOt6vLEHfe/JQGiRRSQQxSfFWpi1MquVdAyjUar5+76PVCmYl" crossorigin="anonymous"></script>
```

## Describe alternatives you've considered

There are other alternative plugins as well. We can also develop an in-house solution.
This feature can also be handled "per app", however, in my opinion, things related to the security of resources loading should be handled by the framework.

**Comment by CodingFlow** [comment]
How do I implement this "per app" for now until it's made part of the framework? My next.config.js:

```javascript
const withPrefresh = require('@prefresh/next')
const SriPlugin = require('webpack-subresource-integrity')

module.exports = withPrefresh({
  webpack(config, { dev, isServer }) {

    // Preact config that is not relevant

    config.output.crossOriginLoading = 'anonymous'
    config.plugins.push(new SriPlugin({
        hashFuncNames: ['sha256', 'sha384'],
        enabled: true,
    }))

    return config
  },
})
```

**Reply by Starefossen** [reply]
@CodingFlow is this example still valid? I should see integrity="…" attributes in the final import in my app after build, right?

**Reply by bmadone** [reply]
@Starefossen did you figure it out?

**Reply by Starefossen** [reply]
> @Starefossen did you figure it out?

Nope!

**Comment by jscheid** [comment]
I'm the principal author of webpack-subresource-integrity, I would also like to see this feature and could maybe help build it.

The snippet above is a start, except it should only be executed when `!isServer`. It will enable SRI for lazy-loaded (dynamically imported) chunks, but of course that doesn't help much when the top-level assets (those loaded from HTML files) don't carry hashes.

Without knowing much about next.js code, it looks like these hashes will have to be extracted from the result of the client build (somewhere around [here](https://github.com/vercel/next.js/blob/42e8dcf281526b5dde27448010e96edcd5ceb196/packages/next/build/index.ts#L382)?) and injected into the HTML pages ([here](https://github.com/vercel/next.js/blob/canary/packages/next/pages/_document.tsx)?).

How would this data be ferried from one place to the other? I've seen [mention of a manifest for this purpose](https://github.com/vercel/next.js/issues/8253#issuecomment-518661266), which should do the trick for the production server. Would that be a good place for it? For static builds the hashes are not needed in a manifest as they will already be baked into the HTML.

I've also seen mention of the client-side router injecting `<script>` tags; these would need treatment as well. It could probably use the hashes that [soon will be exposed on the `__webpack_require__` object](https://github.com/waysact/webpack-subresource-integrity/blob/ef606db9d19949c3a39b14c6bb1fb4969d780123/jmtp.js#L25), assuming that the router feature builds on top of Webpack chunks. Where does that code live?

One other consideration is that when enabled, the [`Cache-Control: no-transform` header should be set](https://github.com/waysact/webpack-subresource-integrity/#proxies). I assume this is straightforward now that next.js has a facility for managing server headers. People who export static pages and host them without TLS might have to be told that they need to ensure this header is set, or opt out of the feature.

**Reply by samuelgoff** [reply]
I'm currently trying to implement a Content Security Policy (CSP), and having `integrity` values added to `script` tags (or the CSP itself) is crucial to making this work well. IMO, having the `integrity` values live inside the `script` tags is vastly more readable than a big blob of hash values living inside the CSP.

I discovered webpack-subresource-integrity and based on its current usage (1.84M downloads per week!), it looks like others have arrived at the same conclusion I have -- that it's a great solution to this problem.

The only concern I have is that I am planning to add TypeScript to my codebase, which means I'll switch over to SWC + Spack, which (as I understand it) replaces babel/webpack.

Is a solution being pursued for SRI in Nextjs? I think it would be low-hanging fruit to significantly improve security of Nextjs apps. If so, will the solution work with the SWC/Spack build, as well?

**Reply by morgansson** [reply]
I'm currently facing this issue, I need SRI and my app is built using Nextjs. 

I believe SRI is the future and the decentralized internet that's ahead of us requires that top frameworks like Nextjs adapt to support this technology. 

I loved the term you've used "low-hanging fruit" to describe how easy would be to implement this. I'm thinking of doing some kind of parser to apply the SRI tags after the build is complete,  as a separate task. I don't know how to actually fix this in any other way with Nextjs 12, as I'm using the swc compiler. Any ideas would be great. 

Thanks a lot to everyone.

**Comment by sambacha** [comment]
does esm modules support SRI even?

**Comment by ajinabraham** [comment]
Any plans on supporting SRI at framework level?

**Comment by Starefossen** [comment]
Does anyone know if the the per-app solution is still working with current version of next.js and webpack? I can not see any integrity attributes being applied for this `next.config.js` when I run `yarn build && yarn start` 😭 

```js
/** @type {import('next').NextConfig} */
const { SubresourceIntegrityPlugin } = require('webpack-subresource-integrity');
const nextConfig = {
  reactStrictMode: true,
  swcMinify: true,
  webpack(config) {
    config.output.crossOriginLoading = "anonymous";
    config.plugins.push(
      new SubresourceIntegrityPlugin({
        hashFuncNames: ["sha256", "sha384"],
        enabled: true,
      })
    );

    return config;
  },
}

module.exports = nextConfig
```

**Comment by praxxis** [comment]
There's been some movement on the SRI front in https://github.com/vercel/next.js/pull/39729, has anyone worked with the new option yet?

**Reply by pkreissel** [reply]
Unfortunately I couldn't find documentation for this. Did you already figure out how to implement this?

**Comment by RodrigoTomeES** [comment]
Hi, Is there any update for pages directory? The plugin `webpack-subresource-integrity`doesn't work anymore 😢

**Reply by bmadone** [reply]
@RodrigoTomeES any updates?

**Reply by RodrigoTomeES** [reply]
@bmadone no, I didn't it work

**Comment by komw** [comment]
Hi, it there any update of this FR?

**Reply by timneutkens** [reply]
There's an experimental feature for SRI:

https://github.com/vercel/next.js/blob/1ce3913ef13a22c03d795df7cc8eec2cf0a7ee37/packages/next/src/server/config-shared.ts#L342

`experimental.sri.algorithm`

**Reply by gutisalex** [reply]
also mentioning this setting only works on webpack...

but what exactly does it do? In my case it does nothing... 

How is it supposed to work? Is there any official documentation on that? Is it the only thing I do in the config? Or do I have to do something else? When can we expect turbo support for it (apparently they are working on it...)?

**Reply by felipedeboni** [reply]
@gutisalex looks like it had a bug, probably it's coming soon now given this [PR](https://github.com/vercel/next.js/pull/39729) was merged.

**Comment by lilong7676** [comment]
I see there is a [canary update](https://github.com/vercel/next.js/pull/73891) that supports SRI on App Router mode，my question is will support Page Route later or not?

**Comment by cbengeWTCT** [comment]
Hi Is there any update on this? We have several clients running Next for their eCommerce stores and this has now become a requirement of PCI-DSS. So would be good to get and answer and some documentation as to how this can be enabled

**Reply by arthurodriguesbatista** [reply]
Same. Did you find any workaround solutions?

**Comment by lance-shuai** [comment]
![image](https://github.com/user-attachments/assets/b3e1f460-11f3-4de6-b4df-dd59a0d3bfc5)

I have built the simplest Next15 project and enabled SRI, but why do some of the simplest projects have script tags without integrity added?

**Reply by lance-shuai** [reply]
![image](https://github.com/user-attachments/assets/c1b2fb08-7941-4f16-8b81-70923b2559f4)

But I did see that all the JS generated hash values in the build product, but I don't know why some script tags were not injected with integrity on the final HTML

**Reply by gergokee** [reply]
i guess this is the reason why it is not added for chunks:
https://github.com/vercel/next.js/blob/38cf78e31efd81cd0fe64c9b3dfaf5290ba60012/test/production/app-dir/subresource-integrity/subresource-integrity.test.ts#L151

**Reply by gergokee** [reply]
you can read this too:
https://github.com/vercel/next.js/pull/73891#issue-2738066595

> In a follow-up we should also add support for adding the integrity attribute to client component chunks that are injected into the head during server-side rendering, but that needs a change in React first.
